base/libgit2/callbacks.jl

```diff
@@ -27,78 +27,162 @@ end
 """Credentials callback function
 
 Function provides different credential acquisition functionality w.r.t. a connection protocol.
-If payload is provided then `payload_ptr` should contain `LibGit2.AbstractPayload` object.
-
-For `LibGit2.Consts.CREDTYPE_USERPASS_PLAINTEXT` type, if payload contains fields: `user` & `pass` they are used to create authentication credentials.
-In addition, if payload has field `used` it can be set to `true` to indicate that payload was used and abort callback with error. This behavior is required to avoid repeated authentication calls with incorrect credentials.
+If payload is provided then `payload_ptr` should contain `LibGit2.AbstractCredentials` object.
+
+For `LibGit2.Consts.CREDTYPE_USERPASS_PLAINTEXT` type, if payload contains fields:
+`user` & `pass`, they are used to create authentication credentials.
+Empty `user` name and `pass`word trigger authentication error.
+
+For `LibGit2.Consts.CREDTYPE_SSH_KEY` type, if payload contains fields:
+`user`, `prvkey`, `pubkey` & `pass`, they are used to create authentication credentials.
+Empty `user` name triggers authentication error.
+
+Order of credential checks (if supported):
+- ssh key pair (ssh-agent if specified in payload's `usesshagent` field)
+- plain text
+
+**Note**: Due to the specifics of `libgit2` authentication procedure, when authentication fails,
+this function is called again without any indication whether authentication was successful or not.
+To avoid an infinite loop from repeatedly using the same faulty credentials,
+`checkused!` function can be called to test if credentials were used if call returns `true` value.
+Used credentials trigger user prompt for (r)entering required information.
+`UserPasswordCredentials` and `CachedCredentials` are implemented a call counting strategy
+that counteracts faulty credentials repeated usage.
```

Comment:
that prevents repeated usage of faulty credentials

(I'm repeating myself, github hid these but they weren't addressed)